fuel-vm/src/interpreter/memory.rs

```diff
@@ -359,6 +343,120 @@ impl MemoryInstance {
     pub fn heap_raw(&self) -> &[u8] {
         &self.heap
     }
+
+    /// Returns a `MemoryRollbackData` that can be used to achieve the state of the
+    /// `desired_memory_state` instance.
+    pub fn collect_rollback_data(
+        &self,
+        desired_memory_state: &MemoryInstance,
+    ) -> Option<MemoryRollbackData> {
+        if self == desired_memory_state {
+            return None
+        }
+
+        let sp = desired_memory_state.stack.len();
+        let hp = desired_memory_state.hp;
+
+        assert!(
+            hp >= self.hp,
+            "We only allow shrinking of the heap during rollback"
+        );
+
+        let stack_changes =
+            get_changes(&self.stack[..sp], &desired_memory_state.stack[..sp], 0);
+
+        let heap_start = hp
```

Comment:
This HP is the one from `desired_memory_state` and we use the `self.heap_offset()` to get heap start point. We should use HP from `self`, no ?

Reply:
The heap can only grow, so `HP` register can only decrease. Here `desired_memory_state` is a memory before we did `grow_heap`, so the `HP` is higher than `self.HP`. We want to subtract `self.heap_offset()` from `desired_memory.HP` to get the offset from which we want to read `self.heap` because `heap` and `desired_heap` should have the same size and should correspond to the same memory location to actually calculate diff.

So the `heap_start` represent part of the memory which we want to skip from the diff, because it is part of the memory allocated by `grow_heap` function